Osiris/Hacks/Misc.cpp

```diff
@@ -207,6 +207,376 @@ static bool worldToScreen(const Vector& in, ImVec2& out) noexcept
+{
+	if (!config->misc.Recorder)
+		return;
+
+	bool isPlaybackActive = playback.IsPlaybackActive();
+	bool isRecordingActive = recorder.IsRecordingActive();
+
+	FrameContainer& recording = recorder.GetActiveRecording();
+
+	bool is_rerecording_active = recorder.IsRerecordingActive();
+	FrameContainer& rerecording = recorder.GetActiveRerecording();
+
+	if (GetAsyncKeyState(0x33)) //Key: 3 Start Recording
+		if (recording.empty())
+			recorder.StartRecording();
+
+	if (GetAsyncKeyState(0x34)) //Key: 4 Stop Recording
+		recorder.StopRecording();
+
+
+	if (GetAsyncKeyState(0x58)) //Key: X Start Playback
+		gotostartpos = true;
+
+	if (gotostartpos)
+	{
+
+		auto dist = localPlayer->getAbsOrigin().distTo(startVec);
+
+		if (cmd->buttons != 0)
+			gotostartpos = false;
+
+		if (dist < 1)
+		{
+			playback.StartPlayback(recorder.GetActiveRecording());
+			gotostartpos = false;
+		}
+
+		else
+		{
+			auto cangle = MCalculateRelativeAngle(localPlayer->getAbsOrigin(), startVec, cmd->viewangles);
+			cmd->viewangles += cangle;
+			cmd->forwardmove = dist;
+		}
+	}
+
+	if (GetAsyncKeyState(0x43)) //Key: C Stop Playback
+		playback.StopPlayback();
+
+	if (GetAsyncKeyState(VK_MENU)) //Key: ALT Clear frames
+		recorder.GetActiveRecording().clear();
+
+	if (GetAsyncKeyState(0x35)) //Key: 5 Save Rerecording
+	{
+		if (recorder.IsRerecordingActive())
+		{
+			recorder.StopRerecording(true);
+			playback.StopPlayback();
+		}
+	}
+
+	if (GetAsyncKeyState(0x36)) //Key: 6 Clear Rerecording
+	{
+		if (recorder.IsRerecordingActive())
+		{
+			recorder.StopRerecording(false);
+			playback.StopPlayback();
+		}
+	}
+
+	if (isRecordingActive)
+	{
+		const size_t current_recording_frame = recording.size();
+
+		movePositions[current_recording_frame] = localPlayer->getAbsOrigin(); //To visualize the movement
+		recording.push_back({ cmd });
+	}
+	else if (is_rerecording_active)
+		rerecording.push_back({ cmd });
+
+	if (isPlaybackActive)
+	{
+		const size_t current_playback_frame = playback.GetCurrentFrame();
+
+		if (cmd->buttons != 0) {
+			recorder.StartRerecording(current_playback_frame);
+			playback.StopPlayback();
+		}
+
+		try
+		{
+			recording.at(current_playback_frame).Replay(cmd);
+			interfaces->engine->setViewAngles(cmd->viewangles); //You can make a config bool for lock the view or not
+
+			if (current_playback_frame + 1 == recording.size())
+			{
+				playback.StopPlayback();
+			}
+			else
+			{
+				playback.SetCurrentFrame(current_playback_frame + 1);
+			}
+		}
+		catch (std::out_of_range)
+		{
+			playback.StopPlayback();
+		}
+	}
+}
+
+void Misc::VisualizeRecorder() noexcept
+{
+	if (config->misc.visualizeRecorder)
+	{
+		/*This isnt working good, its rendering in not good position: https://streamable.com/imlwj8
+		if you have idea how to fix that, please let me know'*/
+
+		ImVec2 pos;
+		ImVec2 pos2;
+
+		FrameContainer& recording = recorder.GetActiveRecording();
+
+		for (int i = 1; i < recording.size(); i++)
```

Comment:
then do movePositions.size() and start from 0

Reply:
thanks